NumLib/ODESolver/NonlinearSolver.cpp

```diff
@@ -232,6 +243,10 @@ NonlinearSolverStatus NonlinearSolver<NonlinearSolverTag::Newton>::solve(
         sys.getJacobian(J);
         INFO("[time] Assembly took %g s.", time_assembly.elapsed());
 
+        // Apply out-of-balance forces if set.
```

Comment:
Probably I'm missing something here, but I think it is not possible to move this part out of the nonlinear solver. The reason is the `sys.getResidual(x, res)` few lines above. Computation of the residuum depends on the time discretization scheme. Passing the time discretization scheme to the processes is not a good option.
The residuum can not be computed in the process in general case.

@norihiro-w @wenqing I'd like to see some sketches/ideas how to move the computation into the processes.

Reply:
@endJunction At least, for mechanics, you don't need to care the time discretization, because it's solving static problems. For dynamic problems, I have no idea how you can make F^oob consistent with any time discretization scheme.